### PR TITLE
Add a test for #14254

### PR DESCRIPTION
**Pull request by Frassle**
Add a test for https://github.com/pulumi/pulumi/issues/14254, which happened to have been "fixed" in https://github.com/pulumi/pulumi/pull/15716. We didn't notice this at the time, and almost regressed this fix in https://github.com/pulumi/pulumi/pull/17245.

Adding a test to ensure we don't regress this.

Fixes https://github.com/pulumi/pulumi/issues/14254.

### CODE REVIEW

File: pkg/engine/lifecycletest/target_test.go [hunk 1]
```diff
@@ -3405,3 +3405,89 @@ func TestUntargetedDependencyChainsArePreserved(t *testing.T) {
...
+		Targets: deploy.NewUrnTargets([]string{
+			"**target**",
+		}),
+	}
+	snap, err = TestOp(Update).RunStep(project, p.GetTarget(t, snap), options, false, p.BackendClient, nil, "1")
+	assert.ErrorContains(t, err,
+		"for resource urn:pulumi:test::test::pkgB:index:typA::unrelated has not been registered yet")
+	// 6 because we have the stack, provider A, target, provider B, unrelated, and the new provider B
+	assert.Equal(t, 6, len(snap.Resources))
+	// unrelated shouldn't have had it's provider changed
```
Comment (lunaris) [inline pkg/engine/lifecycletest/target_test.go:3489]:
```suggestion
	// unrelated shouldn't have had its provider changed
```

File: pkg/engine/lifecycletest/target_test.go [hunk 2]
```diff
@@ -3405,3 +3405,89 @@ func TestUntargetedDependencyChainsArePreserved(t *testing.T) {
 		})
 	})
 }
+
+func TestUntargetedProviderChange(t *testing.T) {
+	// This test is a regression test for https://github.com/pulumi/pulumi/issues/14254. This was "fixed" by
+	// https://github.com/pulumi/pulumi/pull/15716 but we didn't notice. This test is to ensure that the issue
+	// stays fixed because we _almost_ regressed it in https://github.com/pulumi/pulumi/pull/17245.
```
Comment (lunaris) [inline pkg/engine/lifecycletest/target_test.go:3412]: Possible to add a one-liner on what the test is? It's sorta possible to infer it from the comments below but might be useful to have a summary. Also I'd just move this to be the doc-comment for the function.

**Review by lunaris — APPROVED** [review]

File: pkg/engine/lifecycletest/target_test.go (hunk 2 above)
Comment (Frassle) [reply pkg/engine/lifecycletest/target_test.go:3412]: Sure

Comment (Frassle) [reply pkg/engine/lifecycletest/target_test.go]: Comment read ok now?

Comment (lunaris) [reply pkg/engine/lifecycletest/target_test.go]: Awesome :ok_hand: Queued :crossed_fingers:

**Pull request merged**